### PR TITLE
[User management] Use encoded protobuf as page token for listing users. [DPP-888]

### PR DESCRIPTION
**Pull request by pbatko-da**

### CODE REVIEW

**Comment by pbatko-da** [comment]
fyi @rautenrieth-da @meiersi-da

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/page_tokens.proto [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+option java_outer_classname = "PageTokensOuterClass";
+option java_package = "com.daml.ledger.api.v1";
+option csharp_namespace = "Com.Daml.Ledger.Api.V1";
+
+// Describes the payload of a page token for listing users.
+// Not intended to be handled by directly clients and should be presented to them as an opaque string.
+message ListUsersPageTokenPayload {
+    // Describes a page that starts just after `user_id`.
+    string user_id = 1;
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/page_tokens.proto:16]:
```suggestion
    // Users are ordered by ``user_id``, and the next page starts with users whose ``user_id`` is larger than ``user_id_lower_bound_excl``.
    string user_id_lower_bound_excl = 1;
```

File: release/test-protobuf-structure.sh [hunk 2]
```diff
@@ -55,6 +55,7 @@ com/daml/ledger/api/v1/ledger_configuration_service.proto
 com/daml/ledger/api/v1/ledger_identity_service.proto
 com/daml/ledger/api/v1/ledger_offset.proto
 com/daml/ledger/api/v1/package_service.proto
+com/daml/ledger/api/v1/page_tokens.proto
```
Comment (meiersi-da) [inline release/test-protobuf-structure.sh:58]: This should be moved to an `Api` server internal location, as we do not want users to parse the tokens. They should consider them opaque.

@nmarton-da can you advise wrt where to place that file, and code structure overall?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala [hunk 3]
```diff
@@ -259,26 +261,32 @@ object ApiUserManagementService {
...
-              .asGrpcError
-          )
+          .map(_ => invalidToken)
       } yield {
         userId
       }
     }
   }
+
+  private def invalidToken(implicit
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:285]:
```suggestion
  private def invalidPageToken(implicit
```

avoids confusing wrt the different kinds of tokens in the system.

**Review by meiersi-da — COMMENTED** [review]
Looks good up to the placement of the .proto file.

I'm glad we ran this experiment. It's much easier to think about the different kind of page tokens we use as .proto messages.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala (hunk 3 above)
Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:285]: Ok

File: release/test-protobuf-structure.sh (hunk 2 above)
Comment (meiersi-da) [reply release/test-protobuf-structure.sh:58]: We shouldn't use the same namespace as the public ledger api no.

If we go by the content of https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/protobuf/daml/platform/index.proto, then a decent option seems to be:
```
package daml.platform.apiserver;

option java_package = "com.daml.platform.apiserver";
```

No need to define the `csharp_namespace` as this .proto file is internal to our api server.

Comment (meiersi-da) [reply release/test-protobuf-structure.sh:58]: @nmarton-da : what do you think?

Comment (nmarton-da) [reply release/test-protobuf-structure.sh:58]: I agree with @meiersi-da up there, best would be to follow the index.proto in this case:
* no csharp namespace needed
* having the package as apiserver
Only thing is different to index.proto, that this is a API artefact, while index.proto is a persistent one, so probably proto backwards compatibiliy checks will apply here as well. But that is only stronger than what we need here, which is ok until it is not in the way.

File: ledger/participant-integration-api/src/main/protobuf/daml/platform/page_tokens.proto [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+syntax = "proto3";
+
+package daml.platform.apiserver;
+
+option java_package = "com.daml.platform.apiserver";
+
+
+// Describes the payload of a page token for listing users.
+// Not intended to be handled by directly clients and should be presented to them as an opaque string.
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/protobuf/daml/platform/page_tokens.proto:12]:
```suggestion
// Not intended to be handled directly by clients and should be presented to them as an opaque string.
```

File: release/test-protobuf-structure.sh (hunk 2 above)
Comment (pbatko-da) [reply release/test-protobuf-structure.sh]: Ok

**Review by mziolekda — APPROVED** [review]

**Review by meiersi-da — APPROVED** [review]
Thank you @pbatko-da !

**Pull request merged**